### PR TITLE
gh-115041: Add wrappers that are atomic only in free-threaded builds

### PR DESCRIPTION
**Pull request by mpage**
These are intended to be used in places where atomics are required in free-threaded builds, but not in the default build, and we don't want to introduce the potential performance overhead of an atomic operation in the default build.

* Issue: gh-115041

### CODE REVIEW

**Comment by mpage** [comment]
@corona10 @colesbury - Would you please have a look? I ended up using macros after trying experimenting with some of the changes from https://github.com/python/cpython/pull/114843.

**Review by corona10 — COMMENTED** [review]
For the naming, I will delegate to @colesbury 
2 things:
* If the change does not affect to end-user side, we don't add news.d normally :)
* You may need to update [pythoncore.vcxproj](https://github.com/python/cpython/blob/main/PCbuild/pythoncore.vcxproj) and [pythoncore.vcxproj.filters](https://github.com/python/cpython/blob/main/PCbuild/pythoncore.vcxproj.filters) for newly added headers.

**Comment by mpage** [comment]
> If the change does not affect to end-user side, we don't add news.d normally :)

One of the CI signals will fail without either a NEWS entry or the "skip news" label. I haven't been able to figure out how to add the "skip news" label; I suspect that I'm missing some permissions. I don't feel great about asking folks to review changes with a failing CI signal, even if it's benign, so I've been adding a NEWS entry. Do you know if it would be possible to give me (and other Cinder folks who aren't core devs) permission to add "skip news"?

**Comment by corona10** [comment]
> Do you know if it would be possible to give me (and other Cinder folks who aren't core devs) permission to add "skip news"?

You should be promoted to trigger, at least.
But I think that we can discuss about it some moment :) cc @colesbury 
Just feel free to ping me or @colesbury if you need to add the label for it.

**Comment by mpage** [comment]
@colesbury - Would you please take a look at this? This is the last thing we need to sort out before we can merge https://github.com/python/cpython/pull/113830.

**Review by colesbury — COMMENTED** [review]
Thanks for doing this Matt. Looks good, but a few things need to be done:

1) Fix the merge conflict in Makefile.pre.in
2) Add the header to `pythoncore.vcxproj` and `pythoncore.vcxproj.filters`
3) Probably remove the NEWS file based on @corona10's comments

On (2), my strategy is usually to search for references to a nearby header (e.g., `pycore_pyarena.h`) in `PCbuild/` and use the search results to figure out where to add the references to the new header.

File: Makefile.pre.in
Comment (colesbury) [inline Makefile.pre.in]: Should probably remove the added blank line

Comment (mpage) [reply Makefile.pre.in]: Yeah :) I probably should have put this back in draft, but I'm not fully finished tidying this up. I added a dummy commit that actually uses the macros just to make sure things build on windows. Will remove the line and the commit shortly.

**Comment by mpage** [comment]
@colesbury @corona10 - I think this should be good to go now. Would you please take a look?

**Review by colesbury — APPROVED** [review]
This LGTM. @corona10 did you want to look over this as well?

**Review by corona10 — APPROVED** [review]
LGTM!

**Pull request merged**
